### ISSUE TITLE
[ANR] io.bidmachine.* #5570

### ISSUE THREAD

**Issue by raghav2945** [bidmachine/BidMachine-Android-SDK#224]
In our analytical tools, we are receiving a number of ANR from your component. We have not been able to reproduce this problem locally. I'm going to share a few stack traces for you to look at. Please let us know if you require any additional assistance in resolving this issue.

### Stack-trace

BidMachineSharedPreference.java line 54 (io.bidmachine.BidMachineSharedPreference.getInitResponse)
```
main (waiting): tid=1 systid=4736
       at java.lang.Object.wait(Native method)
       at java.lang.Object.wait(Object.java:442)
       at java.lang.Object.wait(Object.java:568)
       at android.app.SharedPreferencesImpl.awaitLoadedLocked(SharedPreferencesImpl.java:279)
       at android.app.SharedPreferencesImpl.contains(SharedPreferencesImpl.java:353)
       at io.bidmachine.BidMachineSharedPreference.getInitResponse(BidMachineSharedPreference.java:54)
       at io.bidmachine.BidMachineImpl.loadStoredInitResponse(BidMachineImpl.java:317)
       at io.bidmachine.BidMachineImpl.initialize(BidMachineImpl.java:146)
       at io.bidmachine.BidMachine.initialize(BidMachine.java:38)
       at io.bidmachine.BidMachine.initialize(BidMachine.java:24)
       at com.audiomack.data.ads.BiddingBidMachineImpl.init(BiddingBidMachine.kt:42)
       at com.audiomack.data.ads.BiddingImpl.init(Bidding.kt:47)
       at com.audiomack.data.ads.AdProvidersHelper.initMoPub$lambda-19(AdProvidersHelper.kt:379)
       at com.audiomack.data.ads.AdProvidersHelper.lambda$GhHC06HJ2khBjePQ3NhBIxdA6Tg(AdProvidersHelper.kt)
       at com.audiomack.data.ads.-$$Lambda$AdProvidersHelper$GhHC06HJ2khBjePQ3NhBIxdA6Tg.onInitializationFinished(lambda)
       at com.mopub.common.MoPub$1.run(MoPub.java:397)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:268)
       at android.app.ActivityThread.main(ActivityThread.java:7963)
       at java.lang.reflect.Method.invoke(Native method)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:627)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:997)
```

BidMachineAd.java line 24 (io.bidmachine.BidMachineAd.access$300)
```
main (runnable): tid=1 systid=4163
       at io.bidmachine.BidMachineAd.access$300(BidMachineAd.java:24)
       at io.bidmachine.BidMachineAd$4.processImpression(BidMachineAd.java:495)
       at io.bidmachine.ViewAdObject$1.onViewTrackingFinished(ViewAdObject.java:88)
       at io.bidmachine.core.VisibilityTracker$TrackingHolder.release(VisibilityTracker.java:119)
       at io.bidmachine.core.VisibilityTracker$TrackingHolder.access$400(VisibilityTracker.java:18)
       at io.bidmachine.core.VisibilityTracker$TrackingHolder$3.run(VisibilityTracker.java:272)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:257)
       at android.app.ActivityThread.main(ActivityThread.java:8192)
       at java.lang.reflect.Method.invoke(Native method)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:626)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1015)
```

Utils.java line 325 (io.bidmachine.core.Utils.getCacheDir)

```
main (runnable): tid=1 systid=10323
       at java.lang.String.hashCode(String.java:1463)
       at android.util.ArrayMap.indexOfKey(ArrayMap.java:434)
       at android.util.ArrayMap.get(ArrayMap.java:484)
       at android.os.ServiceManager.getService(ServiceManager.java:182)
       at android.os.storage.StorageManager.getVolumeList(StorageManager.java:1468)
       at android.os.Environment$UserEnvironment.getExternalDirs(Environment.java:176)
       at android.os.Environment$UserEnvironment.buildExternalStorageAppFilesDirs(Environment.java:301)
       at android.os.Environment.buildExternalStorageAppFilesDirs(Environment.java:1140)
       at android.app.ContextImpl.getExternalFilesDirs(ContextImpl.java:796)
       at android.app.ContextImpl.getExternalFilesDir(ContextImpl.java:789)
       at android.content.ContextWrapper.getExternalFilesDir(ContextWrapper.java:283)
       at io.bidmachine.core.Utils.getCacheDir(Utils.java:325)
       at io.bidmachine.Debugger.setup(Debugger.java:32)
       at io.bidmachine.BidMachineImpl.initialize(BidMachineImpl.java:141)
       at io.bidmachine.BidMachine.initialize(BidMachine.java:38)
       at io.bidmachine.BidMachine.initialize(BidMachine.java:24)
       at com.audiomack.data.ads.BiddingBidMachineImpl.init(BiddingBidMachine.kt:42)
       at com.audiomack.data.ads.BiddingImpl.init(Bidding.kt:47)
       at com.audiomack.data.ads.AdProvidersHelper.initMoPub$lambda-19(AdProvidersHelper.kt:379)
       at com.audiomack.data.ads.AdProvidersHelper.$r8$lambda$ydYqA-A074ewmZwCK4KYErcLTns(AdProvidersHelper.kt)
       at com.audiomack.data.ads.AdProvidersHelper$$ExternalSyntheticLambda33.onInitializationFinished(unavailable)
       at com.mopub.common.MoPub$1.run(MoPub.java:397)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:246)
       at android.app.ActivityThread.main(ActivityThread.java:8595)
       at java.lang.reflect.Method.invoke(Native method)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

**Comment by NikitaCherep** [comment]
@raghav2945 Can you provide more information about this crashes?
1. Android versions
2. Device models
3. BidMachine version
4. Count of crash

**Comment by raghav2945** [comment]
@NikitaCherep, please find the requested details:
```
Android versions : 11, 10, 9, 7
Device models: Samsung A20, A10, A02s; Realme C21; Techno Spark 3, 4,7
BidMachine version: 
	// BidMachine
    implementation 'io.bidmachine:ads:1.7.6'
    implementation 'io.bidmachine:ads.adapters.mopub:1.7.6.25'
Count of ANR: 250+ in last 25 days
```

**Comment by NikitaCherep** [comment]
@raghav2945 Hello! Haven't been able to reproduce yet.

**Comment by raghav2945** [comment]
Yes, I understand, but did you obtain any insight from the shared stack-trace? You may have seen something similar to this stack on some other reporting.

**Comment by NikitaCherep** [comment]
@raghav2945 No one has reported a similar issue.

**Comment by raghav2945** [comment]
No worries, give it a shot and see if you can come up with something. Otherwise, we'll keep an eye on it and, if there are any significant changes, we'll get back to you with more information if we obtain it.

**Comment by NikitaCherep** [comment]
@raghav2945 Hello! We were unable to reproduce this bug, but tried to fix it in version 1.9.2.

Bugs we tried to fix:
- io.bidmachine.BidMachineSharedPreference.getInitResponse
- io.bidmachine.core.Utils.getCacheDir

**Comment by raghav2945** [comment]
> @raghav2945 Hello! We were unable to reproduce this bug, but tried to fix it in version 1.9.2.
> 
> Bugs we tried to fix:
> 
> * io.bidmachine.BidMachineSharedPreference.getInitResponse
> * io.bidmachine.core.Utils.getCacheDir

Thanks @NikitaCherep, for the update.
